### PR TITLE
[mypy] start introducing mypy on lightning_app

### PR DESCRIPTION
**Pull request by nicolai86**
## What does this PR do?

This PR fixes a number of mypy errors which prevent us to enable it for all of `lighning_app` today. To prevent regressions it also enables mypy for the files that were changed.

This is important because it would have caught issues like #14260.

We have 550 more mypy errors to fix before we can enforce mypy for all of lightning_app...

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by awaelchli — COMMENTED** [review]
Great. To properly enforce this, we would need to integrate it with the existing mypy CI job or create one specifically for app, right @Borda?

**Comment by Borda** [comment]
> Great. To properly enforce this, we would need to integrate it with the existing mypy CI job or create one specifically for app, right @Borda?

good point, lest make app and pytorch testing as separate jobs :)

**Review by hhsecond — APPROVED** [review]

**Review by manskx — APPROVED** [review]
Thanks for add that, we defiantly want to add it everywhere :))

maybe you could rebase your PR ?

**Comment by nicolai86** [comment]
rebased!

File: pyproject.toml [hunk 1]
```diff
@@ -26,8 +26,14 @@ exclude = '(_notebooks/.*)'
 [tool.mypy]
 files = [
     "src/pytorch_lightning",
+    "src/lightning_app/cli/lightning_cli*.py",
+    "src/lightning_app/cli/cmd_*.py",
```
Comment (Borda) [inline pyproject.toml:30]: can we rather ad testing all app folders and list excluded below in modules?
so it is consistent with pytorch typing check setting

Comment (nicolai86) [reply pyproject.toml:30]: let me resolve the merge conflicts first and then change the pyproject.toml..

**Comment by Borda** [comment]
@nicolai86, mind checking the collisions? :otter:

File: pyproject.toml (hunk 1 above)
Comment (nicolai86) [reply pyproject.toml]: resolved and followed your suggestions @Borda

File: pyproject.toml [hunk 2]
```diff
@@ -53,6 +59,86 @@ warn_no_return = "False"
 module = [
     "pytorch_lightning.callbacks.progress.rich_progress",
     "pytorch_lightning.trainer.trainer",
-    "pytorch_lightning.tuner.batch_size_scaling"
+    "pytorch_lightning.tuner.batch_size_scaling",
```
Comment (Borda) [inline pyproject.toml:62]: Perfect! thank you :otter:

**Review by Borda — APPROVED** [review]

File: pyproject.toml [hunk 3]
```diff
@@ -53,6 +59,86 @@ warn_no_return = "False"
...
+    "lightning_app.api.http_methods",
```
Comment (awaelchli) [inline pyproject.toml:63]: That's a long list 😅 
And it took us 1y+ to get the ones in PL figured out. 
Good luck to us 🤞

File: src/lightning_app/cli/cmd_pl_init.py [hunk 4]
```diff
@@ -121,7 +121,7 @@ def project_file_from_template(template_dir: Path, destination_dir: Path, templa
 def print_pretty_report(
     directory: pathlib.Path,
     ignore_patterns: Optional[List[str]] = None,
-    help_texts: Optional[Dict[str, str]] = None,
+    help_texts: Dict[str, str] = {},
```
Comment (awaelchli) [inline src/lightning_app/cli/cmd_pl_init.py:124]: Let's revert this, and in the body do

`help_texts = {} if help_texts is None else help_texts`

**Review by awaelchli — APPROVED** [review]

**Comment by Borda** [comment]
@nicolai86 one more item to fix: 

`src/lightning_app/runners/backends/__init__.py:18: error: Cannot instantiate abstract class "DockerBackend" with abstract attributes "resolve_url" and "stop_work"  [abstract]`

File: pyproject.toml (hunk 3 above)
Comment (carmocca) [reply pyproject.toml:63]: And a huge help from the community lol

**Comment by nicolai86** [comment]
> @nicolai86 one more item to fix:
> 
> `src/lightning_app/runners/backends/__init__.py:18: error: Cannot instantiate abstract class "DockerBackend" with abstract attributes "resolve_url" and "stop_work" [abstract]`

I saw that there was 1 error yesterday but no output on the mypy workflow. let me fix this :+1:

File: src/lightning_app/cli/cmd_pl_init.py (hunk 4 above)
Comment (nicolai86) [reply src/lightning_app/cli/cmd_pl_init.py]: reverted

File: pyproject.toml (hunk 3 above)
Comment (nicolai86) [reply pyproject.toml:62]: we'll get there :+1:

**Pull request merged**